modules/imgcodecs/src/grfmt_jpeg2000_openjpeg.cpp

```diff
@@ -0,0 +1,694 @@
+    // FIXME: raw code stream
+}
+
+
+ImageDecoder Jpeg2KOpjDecoder::newDecoder() const
+{
+    return makePtr<Jpeg2KOpjDecoder>();
+}
+
+bool Jpeg2KOpjDecoder::readHeader()
+{
+    stream_.reset(opj_stream_create_default_file_stream(m_filename.c_str(), OPJ_STREAM_READ));
+    if (!stream_)
+        return false;
+
+    codec_.reset(opj_create_decompress(OPJ_CODEC_JP2));
+    if (!codec_)
+        return false;
+
+    // Callbacks are cleared, when opj_destroy_codec is called,
+    // They can provide some additional information for the user, about what goes wrong
+    setupLogCallbacks(codec_.get());
+
+    opj_dparameters parameters = setupDecoderParameters();
+    if (!opj_setup_decoder(codec_.get(), &parameters))
+        return false;
+
+    {
+        opj_image_t* rawImage;
+        if (!opj_read_header(stream_.get(), codec_.get(), &rawImage))
+            return false;
+
+        image_.reset(rawImage);
+    }
+
+    m_width = image_->x1 - image_->x0;
+    m_height = image_->y1 - image_->y0;
+
+    /* Different components may have different precision,
+     * so check all.
+     */
+    bool hasAlpha = false;
+    const int numcomps = image_->numcomps;
+    CV_Assert(numcomps >= 1);
+    for (int i = 0; i < numcomps; i++)
+    {
+        const opj_image_comp_t& comp = image_->comps[i];
+
+        if (comp.sgnd)
+        {
+            CV_Error(Error::StsNotImplemented, cv::format("OpenJPEG2000: Component %d/%d is signed", i, numcomps));
+        }
+
+        if (hasAlpha && comp.alpha)
+        {
+            CV_Error(Error::StsNotImplemented, cv::format("OpenJPEG2000: Component %d/%d is duplicate alpha channel", i, numcomps));
+        }
+
+        hasAlpha |= comp.alpha;
+
+        m_maxPrec = std::max(m_maxPrec, comp.prec);
+    }
+
+    if (m_maxPrec < 8) {
+        CV_Error(Error::StsNotImplemented, "OpenJPEG2000: Precision < 8 not supported");
+    } else if (m_maxPrec == 8) {
+        m_type = CV_MAKETYPE(CV_8U, numcomps);
+    } else if (m_maxPrec <= 16) {
+        m_type = CV_MAKETYPE(CV_16U, numcomps);
+    } else if (m_maxPrec <= 23) {
+        m_type = CV_MAKETYPE(CV_32F, numcomps);
+    } else {
+        m_type = CV_MAKETYPE(CV_64F, numcomps);
+    }
+    return true;
+}
+
+bool Jpeg2KOpjDecoder::readData( Mat& img )
+{
+    using DecodeFunc = bool(*)(const opj_image_t&, cv::Mat&, uint8_t shift);
+
+    if (!opj_decode(codec_.get(), stream_.get(), image_.get()))
+    {
+        CV_Error(Error::StsError, "OpenJPEG2000: Decoding is failed");
+    }
+
+    if (img.channels() == 2)
+    {
+        CV_Error(Error::StsNotImplemented,
+                 cv::format("OpenJPEG2000: Unsupported number of output channels. IN: %d OUT: 2", image_->numcomps));
+    }
+
+    DecodeFunc decode = nullptr;
+    switch (image_->color_space)
+    {
+    case OPJ_CLRSPC_UNKNOWN:
+        CV_LOG_WARNING(nullptr, "OpenJPEG2000: Image has unknown color space, SRGB is assumed");
+    case OPJ_CLRSPC_SRGB:
```

Comment:
Add `falltru` comment to avoid compiler warnings:

```
        CV_LOG_WARNING(nullptr, "OpenJPEG2000: Image has unknown color space, SRGB is assumed");
        // fallthru
    case OPJ_CLRSPC_SRGB:
```